apollo-router/src/services/subgraph_service.rs

```diff
@@ -422,6 +439,133 @@ impl tower::Service<SubgraphRequest> for SubgraphService {
+            mode = %"sse",
+        );
+
+        // Dedup happens here
+        return Ok(SubgraphResponse::builder()
+            .context(context)
+            .extensions(Object::default())
+            .build());
+    }
+
+    let (parts, body) = subgraph_request.into_parts();
+
+    let mut builder = get_sse_client(service_name.clone(), parts, subgraph_cfg)?;
+    let body = serde_json::to_string(&body)?;
+    builder = builder.body(body);
+
+    let display_headers = context.contains_key(LOGGING_DISPLAY_HEADERS);
+    let display_body = context.contains_key(LOGGING_DISPLAY_BODY);
+    if display_headers {
+        tracing::info!(http.request.headers = ?builder.get_headers(), apollo.subgraph.name = %service_name, "SSE request headers to subgraph {service_name:?}");
+    }
+    if display_body {
+        tracing::info!(http.request.body = ?builder.get_body().unwrap_or_default(), apollo.subgraph.name = %service_name, "SSE request body to subgraph {service_name:?}");
+    }
+
+    let uri = builder.get_url();
+    let path = uri.path();
+    let host = uri.host().unwrap_or_default();
+    let port = uri.port_u16().unwrap_or_else(|| {
+        let scheme = uri.scheme_str();
+        if scheme == Some("https") {
+            443
+        } else if scheme == Some("http") {
+            80
+        } else {
+            0
+        }
+    });
+
+    let subgraph_req_span = tracing::info_span!("subgraph_request",
+        "otel.kind" = "CLIENT",
+        "net.peer.name" = %host,
+        "net.peer.port" = %port,
+        "http.route" = %path,
+        "http.url" = %uri,
+        "net.transport" = "ip_tcp",
+        "apollo.subgraph.name" = %service_name,
+        "graphql.operation.name" = %operation_name,
+    );
+
+    get_text_map_propagator(|propagator| {
+        propagator.inject_context(
+            &subgraph_req_span.context(),
+            &mut opentelemetry_http::HeaderInjector(builder.headers_mut()),
+        );
+    });
+
+    let gql_stream = GraphqlSSE::new(
+        convert_sse_stream(builder.build(), subscription_hash.clone()),
+        subscription_hash,
+    )
+    .map_err(|err| FetchError::SubrequestWsError {
+        service: service_name,
+        reason: format!("cannot send the subgraph request to sse stream: {err:?}"),
+    })?;
+
+    let (mut handle_sink, handle_stream) = handle.split();
+
+    tokio::task::spawn(async move {
+        let mut stream = gql_stream.map(Ok::<_, graphql::Error>);
+
+        while let Some(val) = stream.next().await {
+            if let Ok(val) = val {
+                if let Err(err) = handle_sink.send(val).await {
+                    tracing::trace!(
+                        "cannot send the sse stream to the subscription stream: {err:?}"
+                    );
+                    break;
+                }
+            }
+        }
+
+        if let Err(err) = handle_sink.close().await {
+            tracing::trace!("cannot close the websocket stream: {err:?}");
+        }
```

Comment:
Why don't you use the `forward` method as used in `call_websocket` ?

Reply:
I tried with `forward`, but `forward` will keep waiting when the sink is closed and will only finish when the stream finishes